bundles/org.openhab.core.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/AutomaticInboxProcessor.java

```diff
@@ -92,19 +94,33 @@
 
     private final Logger logger = LoggerFactory.getLogger(this.getClass());
 
-    @NonNullByDefault({})
-    private ThingRegistry thingRegistry;
-    @NonNullByDefault({})
-    private ThingTypeRegistry thingTypeRegistry;
-    @NonNullByDefault({})
-    private Inbox inbox;
+    private final ThingRegistry thingRegistry;
+    private final ThingTypeRegistry thingTypeRegistry;
+    private final Inbox inbox;
     private boolean autoIgnore = true;
     private boolean alwaysAutoApprove = false;
 
     private final Set<InboxAutoApprovePredicate> inboxAutoApprovePredicates = new CopyOnWriteArraySet<>();
 
-    public AutomaticInboxProcessor() {
+    @Activate
+    public AutomaticInboxProcessor(final @Reference ThingTypeRegistry thingTypeRegistry,
+            final @Reference ThingRegistry thingRegistry, final @Reference Inbox inbox) {
         super(ThingStatusInfoChangedEvent.TYPE);
+        this.thingTypeRegistry = thingTypeRegistry;
+        this.thingRegistry = thingRegistry;
+        this.inbox = inbox;
+
+        // This should be the last step (to be more precise: providing the "this" reference to other ones as long as
```

Comment:
Hm, I do not understand from this comment, why these two lines couldn't be moved to the `activate()` method, which would seem to be the much better place to me.

Reply:
Yes, we can do that. But maybe I understood it wrong: Isn't the constructor part of the activation in this case?

Reply:
As you added the `@Activate` annotation to the constructor, the constructor "is" the activate method.

Reply:
Thus I conclude that the `@Activate` annotation is not mandatory for constructor injection? If so @kaikreuzer proposal sounds right.

Reply:
> the constructor "is" the activate method

I learned something new - thanks 🤗

Reply:
Sorry, you understand my comment not correctly.
My comment has been about the code that is used by @cweitkamp (so, the reviewed one).

Here my understanding of the specification:
If constructor injection is used, the constructor must be annotated with `@Activate`.
You can add another activation method then you need to add another method that is annotated by `@Activate`.

There is an example here:
https://osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#service.component-reference.element

So, the preferred solution here and for the other (IIRC it has been persistence inbox) would be to use the constructor injection (activate annotated constructor) and another activate method (activate annotated method).

``java
    @Activate
    public AutomaticInboxProcessor(final @Reference ThingTypeRegistry thingTypeRegistry,
            final @Reference ThingRegistry thingRegistry, final @Reference Inbox inbox) {
        super(ThingStatusInfoChangedEvent.TYPE);
        this.thingTypeRegistry = thingTypeRegistry;
        this.thingRegistry = thingRegistry;
        this.inbox = inbox;
    }

    @Activate
    public void activate() {
        this.thingRegistry.addRegistryChangeListener(this);
        this.inbox.addInboxListener(this);
    }
```

Reply:
Hm, that is pretty confusing that the constructor requires the annotation as well, although it does not serve as the activation method... So then it would indeed be cleaner to add a dedicated `activate()` method for the listeners.